src/main/java/org/quantumbadger/redreader/reddit/prepared/html/HtmlRawElementImg.java

```diff
@@ -0,0 +1,169 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * RedReader is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with RedReader.  If not, see <http://www.gnu.org/licenses/>.
+ ******************************************************************************/
+
+package org.quantumbadger.redreader.reddit.prepared.html;
+
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+import android.text.Spannable;
+import android.text.SpannableStringBuilder;
+import android.text.style.ImageSpan;
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.appcompat.app.AppCompatActivity;
+
+import org.quantumbadger.redreader.account.RedditAccountManager;
+import org.quantumbadger.redreader.cache.CacheManager;
+import org.quantumbadger.redreader.cache.CacheRequest;
+import org.quantumbadger.redreader.cache.CacheRequestCallbacks;
+import org.quantumbadger.redreader.cache.downloadstrategy.DownloadStrategyIfNotCached;
+import org.quantumbadger.redreader.common.Constants;
+import org.quantumbadger.redreader.common.General;
+import org.quantumbadger.redreader.common.GenericFactory;
+import org.quantumbadger.redreader.common.Optional;
+import org.quantumbadger.redreader.common.Priority;
+import org.quantumbadger.redreader.common.datastream.SeekableInputStream;
+import org.quantumbadger.redreader.http.FailedRequestBody;
+import org.quantumbadger.redreader.reddit.prepared.bodytext.BodyElement;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+
+public class HtmlRawElementImg extends HtmlRawElement{
+	@NonNull private static final String TAG = "HtmlRawElementImg";
+
+	@NonNull private final ArrayList<HtmlRawElement> mChildren;
+	@NonNull private final String mTitle;
+	@NonNull private final String mSrc;
+
+	public HtmlRawElementImg(
+			@NonNull final ArrayList<HtmlRawElement> children,
+			@NonNull final String title,
+			@NonNull final String src) {
+		mChildren = children;
+		mTitle = title;
+		mSrc = src;
+	}
+
+	@Override
+	public void getPlainText(@NonNull final StringBuilder stringBuilder) {
+		for(final HtmlRawElement element : mChildren) {
+			element.getPlainText(stringBuilder);
+		}
+	}
+
+	public final synchronized void writeTo(
+			@NonNull final SpannableStringBuilder ssb,
+			@NonNull final AppCompatActivity activity) {
+		final int emoteLocationStart = ssb.length();
+
+		final CountDownLatch latch = new CountDownLatch(1);
```

Comment:
It looks like this is blocking the thread until the download has completed. This is something we need to avoid, because:

* Some HTML parsing is done on the main thread, and if this blocks, then it could cause UI stuttering or ANRs (crashes)
* A comment section with lots of emotes would download these one at a time, so if for example we have 100 emotes, we'd need to block for 100 round-trip network requests before we show the comments, which could take several minutes on poor connections

Ideally we need to perform the downloads asynchronously in the background, in such a way that we can display the comments without waiting for the downloads to complete. If a download completes after the comment containing the emote is already displayed, then the comment should be updated.

This will require a bit of refactoring -- I'd suggest adding an "invalidate" callback as a parameter to `generate()`.

```java
public void generate(
			@NonNull final AppCompatActivity activity,
			@NonNull final ArrayList<BodyElement> destination,
			@NonNull final Runnable invalidateCallback)
```

Once the image was downloaded, `HtmlRawElementImg` would:

* Store the image in an `AtomicReference` member variable
* Invoke the invalidate callback

The code responsible for rendering the comment would then know to call `generate()` again. This time, the `HtmlRawElementImg` would already have the image in memory.

There would also need to be some changes to `HtmlReader.parse()` -- currently this:

* Parses the HTML
* Produces a reduced `HtmlRawElementBlock` by calling `reduce()` on the tree
* Produces a `BodyElementVerticalSequence` by calling `generate()` on the reduced block

That last step needs to be moved outside `parse()`, so that it can be run again (when the invalidate callback is invoked) without the need to do the parsing and reduction again (which would mean a new instance of `HtmlRawElementImg` would be created, and it wouldn't have the image cached in memory).

Hopefully that all makes sense! Happy to answer any questions about this.

Reply:
Thanks for the feedback! I'll look into `HtmlReader.parse()` function when I have bit of time and see if I can make sense of your directions. 

I have just one question for now. Wouldn't it be simpler and more efficient to just update the textview in the callback?

Reply:
Thanks! Updating the `TextView` is probably fine too, it depends on exactly how it's implemented.

Feel free to implement it however you think is best and I'll take a look :)

Reply:
I made the emote download non-blocking. 

I created a callback in `BodyElementTextSpanned` that would update the `TextView` content based on the content of the `AtomicReference` variable. 

Not sure if this is the best way to do this or if updating the `TextView` could have side effects. Would appreciate your feedback